sklearn/cluster/_spectral.py

```diff
@@ -18,6 +18,40 @@
 from ._kmeans import k_means
 
 
+def cluster_qr(vectors):
+    """Search for a partition matrix (clustering) which is
+    closest to the eigenvector embedding.
+
+    This implementation was proposed in [1]_.
+
+    Parameters
+    ----------
+    vectors : array-like, shape: (n_samples, n_clusters)
+        The embedding space of the samples.
+
+    Returns
+    -------
+    labels : array of integers, shape: n_samples
+        The labels of the clusters.
+
+    References
+    ----------
+    .. [1] `Simple, direct, and efficient multi-way spectral clustering, 2019
+        Anil Damle, Victor Minden, Lexing Ying
+        <https://doi.org/10.1093/imaiai/iay008>`_
+
+    """
+
+    from scipy.linalg import qr, svd
+
+    k = vectors.shape[1]
+    piv = qr(vectors.T, pivoting=True)[2]
+    piv = piv[0:k]
+    ut, _, v = svd(vectors[piv[:k], :].T)
+    vectors = abs(np.dot(vectors, np.dot(ut, v.conj())))
```

Comment:
Would https://numpy.org/doc/stable/reference/generated/numpy.linalg.multi_dot.html help find the optimal order of the matrix multiplications based on the shapes? Or is this order always optimal?